### PR TITLE
#6540 Feature: Adds TikTok Provider

### PR DESCRIPTION
**Pull request by sdairs**
## 🧢 Checklist

- [x] Documentation
- [ ] Tests
- [ ] Ready to be merged

## 🎫 Affected issues

https://github.com/nextauthjs/next-auth/issues/6540

### CODE REVIEW

**Comment by balazsorban44** [comment]
Actually, I realize we already had: https://github.com/nextauthjs/next-auth/pull/2720, but could not merge because I was unable to test. I wanted to create an application, but TikTok hasn't approved it. <picture data-single-emoji=":sadblob:" title=":sadblob:">:sadblob:</picture> Do you have some suggestions on how to get an app approved? :pray: Or if you have an already-approved dev app I could loan to test this out, you could reach out to me on Twitter maybe.

File: packages/core/src/providers/tiktok.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    name: "TikTok",
+    type: "oauth",
+    version: "2.0",
+    authorization: {
+      url: "https://www.tiktok.com/auth/authorize/",
+      params: {
+        scope: "user.info.basic",
+        response_type: "code",
+        client_key: options.clientId,
+        redirect_uri: options.redirect_uri,
```
Comment (balazsorban44) [inline packages/core/src/providers/tiktok.js:14]:
```suggestion
        client_key: options.clientId,
```

already handled

File: packages/core/src/providers/tiktok.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/** @type {import(".").OAuthProvider} */
+export default function TikTok(options) {
+  return {
+    id: "tiktok",
+    name: "TikTok",
+    type: "oauth",
+    version: "2.0",
+    authorization: {
+      url: "https://www.tiktok.com/auth/authorize/",
```
Comment (balazsorban44) [inline packages/core/src/providers/tiktok.js:9]:
```suggestion
      url: "https://www.tiktok.com/auth/authorize",
```

File: packages/core/src/providers/tiktok.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        client_key: options.clientId,
+        redirect_uri: options.redirect_uri,
+      },
+    },
+    token: {
+      url: "https://open-api.tiktok.com/oauth/access_token/",
+      params: {
+        client_key: options.clientId,
+        client_secret: options.clientSecret,
+        grant_type: "authorization_code",
```
Comment (balazsorban44) [inline packages/core/src/providers/tiktok.js:22]:
```suggestion
      url: "https://open-api.tiktok.com/oauth/access_token",
      params: {
        client_key: options.clientId,
        client_secret: options.clientSecret,
```

already handled

File: packages/core/src/providers/tiktok.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    },
+    userinfo: "https://open-api.tiktok.com/user/info/",
+    profile(profile) {
+      return {
+        id: profile.open_id,
+        name: profile.display_name,
+        image: profile.avatar_url
+      };
+    },
+    checks: ["state"],
```
Comment (balazsorban44) [inline packages/core/src/providers/tiktok.js:33]:
```suggestion
```

File: packages/core/src/providers/tiktok.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        client_secret: options.clientSecret,
+        grant_type: "authorization_code",
+      },
+    },
+    userinfo: "https://open-api.tiktok.com/user/info/",
+    profile(profile) {
+      return {
+        id: profile.open_id,
+        name: profile.display_name,
+        image: profile.avatar_url
```
Comment (balazsorban44) [inline packages/core/src/providers/tiktok.js:30]:
```suggestion
        image: profile.avatar_url,
        email: profile.email ?? null
```

Do you know if e-mail can be returned?

File: packages/core/src/providers/tiktok.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      },
+    },
+    token: {
+      url: "https://open-api.tiktok.com/oauth/access_token",
+      params: {
+        client_key: options.clientId,
+        client_secret: options.clientSecret,
+      },
+    },
+    userinfo: "https://open-api.tiktok.com/user/info/",
```
Comment (balazsorban44) [inline packages/core/src/providers/tiktok.js:22]:
```suggestion
    userinfo: "https://open-api.tiktok.com/user/info",
```

File: packages/core/src/providers/tiktok.js (hunk 5 above)
Comment (sdairs) [reply packages/core/src/providers/tiktok.js:27]: https://developers.tiktok.com/doc/tiktok-api-v2-get-user-info/  if you scroll down on this page, there is docs for the User Object - I couldn't find any ref to get an email unfortunately

**Comment by sdairs** [comment]
> 

Yeah it was a nightmare to get an app approved by TikTok. I have one, I'll DM and we can use it to test

File: packages/core/src/providers/tiktok.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      },
+    },
+    token: {
+      url: "https://open-api.tiktok.com/oauth/access_token",
+      params: {
+        client_key: options.clientId,
+        client_secret: options.clientSecret,
+      },
+    },
+    userinfo: "https://open-api.tiktok.com/user/info",
```
Comment (maiconcarraro) [inline packages/core/src/providers/tiktok.js:22]: why not already use v2? https://developers.tiktok.com/doc/tiktok-api-v2-get-user-info/

**Comment by angelhodar** [comment]
Hey everyone! If you need some help with this PR I can give a helping hand here. I already have a tiktok aopp that I used for testing, if you want I can change the redirect URL to some testing domain (localhost is not allowed) @balazsorban44

**Comment by kaldray** [comment]
Hey everyone what the actual state of this PR ?

**Comment by ThangHuuVu** [comment]
close via #8131

**Pull request closed without merging**